File: python/src/main/python/Q049.py
from src.main.python.Solution import Solution

# Given an array of strings, group anagrams together.
#
# For example, given: ["eat", "tea", "tan", "ate", "nat", "bat"],
#   Return:
#
#   [
#     ["ate", "eat","tea"],
#     ["nat","tan"],
#     ["bat"]
#   ]

# Note:
#   For the return value, each inner list's elements must follow the lexicographic order.
#   All inputs will be in lower-case.


class Q049(Solution):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        ans = []
        if strs:
            table = {}
            for str in strs:
                key = ''.join(sorted(str))
                anagrams = table.get(key, [])
                anagrams.append(str)
                table[key] = anagrams
            for key in table:
                ans.append(sorted(table[key]))
        return ans

File: java/src/main/java/solutions/Q049.java
package solutions;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Given an array of strings, group anagrams together.
 *
 * For example, given: ["eat", "tea", "tan", "ate", "nat", "bat"],
 *
 * Return:
 *  [
 *      ["ate", "eat","tea"],
 *      ["nat","tan"],
 *      ["bat"]
 *  ]
 */

public class Q049 extends Solution {

    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> ans = new ArrayList<>();
        if (strs == null || strs.length == 0) return ans;
        Map<String, List<String>> hm = new HashMap<>();
        for (String s : strs) {
            char[] ch = s.toCharArray();
            Arrays.sort(ch);
            String key = new String(ch);
            hm.putIfAbsent(key, new ArrayList<>());
            hm.get(key).add(s);
        }
        ans.addAll(hm.values());
        return ans;
    }
}
